jbmc/src/java_bytecode/remove_exceptions.cpp

```diff
@@ -472,7 +478,21 @@ void remove_exceptionst::instrument_exceptions(
 
   Forall_goto_program_instructions(instr_it, goto_program)
   {
-    if(instr_it->is_decl())
+    if(instr_it->is_assert())
+    {
+      if(propagate_assertion_error)
+      {
+        // suppress user-provided assertion
+        // because we propgate the AssertionError
+        if(
+          instr_it->guard.is_false() &&
+          instr_it->source_location.get_bool("user-provided"))
+        {
+          instr_it->make_skip();
```

Comment:
Instead of fixing this up afterwards, how about doing it right the first time? This gets handled here (https://github.com/diffblue/cbmc/blob/10a468514c5b793f935e19029b66437586bedfe0/src/goto-programs/builtin_functions.cpp#L740), but this suggests the frontend should be in charge of adding the `ASSERT(FALSE)` or not as appropriate

Reply:
Yes, but goto-programs is not the right place for adding more java-specifics and needs to be refactored.

Reply:
Right, that's what I'm saying. Instead of doing something weird here and then later `goto_convertt` special-cases `java::java.lang.AssertionError.<init>:`, why not remove the special case and have the frontend / java-instrument pass do the right thing (either generate an `ASSERT FALSE` or don't, depending on whether the user wants `AssertionError` to be propagated)